### PR TITLE
[ntuple] Fixes for 32bit / non-x86_64 builds

### LINKED ISSUE #12427: Test failures in RNTuple: 32 bit specific
- [x] Checked for duplicates

### Describe the bug

Two RNTuple tests fail on ix86 (Intel 32 bit).

### Expected behavior

The tests should pass.

### To Reproduce

1. Compile root 6.28.00 for ix86.
2. Run unit tests.

### Setup

1. ROOT version: 6.28.00
2. Operating system: Fedora Linux 36/37/38/39 on i686
3. How you obtained ROOT: compilation from source.

### Additional context

Logs for the two failing tests:

gtest-tree-ntuple-v7-test-ntuple-types fails due to alignment differences for std::tuple and std::pair.

~~~
 479/1318 Test  #444: gtest-tree-ntuple-v7-test-ntuple-types ..............................***Failed   13.57 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet:RNTuple.TClassEBO
[==========] Running 18 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 18 tests from RNTuple
[ RUN      ] RNTuple.StdPair
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_types.cxx:102: Failure
Expected equality of these values:
  (alignof(std::pair<int64_t, float>))
    Which is: 4
  otherField->GetAlignment()
    Which is: 8
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[  FAILED  ] RNTuple.StdPair (287 ms)
[ --- ]
[ RUN      ] RNTuple.StdTuple
/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_types.cxx:151: Failure
Expected equality of these values:
  (alignof(std::tuple<char, int64_t, char>))
    Which is: 4
  otherField->GetAlignment()
    Which is: 8
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[  FAILED  ] RNTuple.StdTuple (317 ms)
[ --- ]
[----------] 18 tests from RNTuple (7848 ms total)
[----------] Global test environment tear-down
[==========] 18 tests from 1 test suite ran. (7848 ms total)
[  PASSED  ] 16 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] RNTuple.StdPair
[  FAILED  ] RNTuple.StdTuple
 2 FAILED TESTS

~~~

gtest-tree-ntuple-v7-test-ntuple-extended fails with:
Fatal: fWritePage[otherIdx].IsEmpty() violated

~~~
 520/1318 Test  #454: gtest-tree-ntuple-v7-test-ntuple-extended ...........................***Failed   73.15 sec
Running main() from /builddir/build/BUILD/googletest-1.13.0/googletest/src/gtest_main.cc
Note: Google Test filter = -RCsvDS.Remote:RRawFile.Remote:RSqliteDS.Davix:TFile.ReadWithoutGlobalRegistrationWeb:TFile.ReadWithoutGlobalRegistrationNet
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from RNTuple
[ RUN      ] RNTuple.RealWorld1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.RealWorld1 (9118 ms)
[ RUN      ] RNTuple.RandomAccess
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.RandomAccess (201 ms)
[ RUN      ] RNTuple.LargeFile1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTupleSerialize.cxx:1148 in static ROOT::Experimental::RResult<void> ROOT::Experimental::Internal::RNTupleSerializer::DeserializeHeaderV1(const void*, uint32_t, ROOT::Experimental::RNTupleDescriptorBuilder&)>: Pre-release format version: RC 1
[       OK ] RNTuple.LargeFile1 (34858 ms)
[ RUN      ] RNTuple.LargeFile2
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Warning in <[ROOT.NTuple] Warning /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RPageStorageFile.cxx:51 in ROOT::Experimental::Detail::RPageSinkFile::RPageSinkFile(std::string_view, const ROOT::Experimental::RNTupleWriteOptions&)>: The RNTuple file format will change. Do not store real data with this version of RNTuple!
Fatal: fWritePage[otherIdx].IsEmpty() violated at line 82 of `/builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RColumn.cxx'
aborting
Thread 6 (Thread 0xe26ffb40 (LWP 117258) "ntuple_extended"):
#0  0xf7ef1559 in __kernel_vsyscall ()
#1  0xf69239bb in syscall () from /lib/libc.so.6
#2  0xf7338c73 in tbb::internal::rml::private_worker::thread_routine(void*) () from /lib/libtbb.so.2
#3  0xf689f47d in start_thread () from /lib/libc.so.6
#4  0xf6927d6c in clone3 () from /lib/libc.so.6
Thread 5 (Thread 0xe2afeb40 (LWP 117257) "ntuple_extended"):
#0  0xf7ef1559 in __kernel_vsyscall ()
#1  0xf69239bb in syscall () from /lib/libc.so.6
#2  0xf7338c73 in tbb::internal::rml::private_worker::thread_routine(void*) () from /lib/libtbb.so.2
#3  0xf689f47d in start_thread () from /lib/libc.so.6
#4  0xf6927d6c in clone3 () from /lib/libc.so.6
Thread 4 (Thread 0xe2cffb40 (LWP 117256) "ntuple_extended"):
#0  0xf7ef1559 in __kernel_vsyscall ()
#1  0xf69239bb in syscall () from /lib/libc.so.6
#2  0xf7338c73 in tbb::internal::rml::private_worker::thread_routine(void*) () from /lib/libtbb.so.2
#3  0xf689f47d in start_thread () from /lib/libc.so.6
#4  0xf6927d6c in clone3 () from /lib/libc.so.6
Thread 3 (Thread 0xe300cb40 (LWP 117255) "ntuple_extended"):
#0  0xf7ef1559 in __kernel_vsyscall ()
#1  0xf69239bb in syscall () from /lib/libc.so.6
#2  0xf7338c73 in tbb::internal::rml::private_worker::thread_routine(void*) () from /lib/libtbb.so.2
#3  0xf689f47d in start_thread () from /lib/libc.so.6
#4  0xf6927d6c in clone3 () from /lib/libc.so.6
Thread 2 (Thread 0xe320db40 (LWP 117254) "ntuple_extended"):
#0  0xf7ef1559 in __kernel_vsyscall ()
#1  0xf69239bb in syscall () from /lib/libc.so.6
#2  0xf7338c73 in tbb::internal::rml::private_worker::thread_routine(void*) () from /lib/libtbb.so.2
#3  0xf689f47d in start_thread () from /lib/libc.so.6
#4  0xf6927d6c in clone3 () from /lib/libc.so.6
Thread 1 (Thread 0xf73179c0 (LWP 117207) "ntuple_extended"):
#0  0xf7ef1559 in __kernel_vsyscall ()
#1  0xf68e3e1a in __wait4_time64 () from /lib/libc.so.6
#2  0xf68e3c3a in waitpid () from /lib/libc.so.6
#3  0xf6862810 in do_system () from /lib/libc.so.6
#4  0xf70c9ebb in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x565f30c0) at /builddir/build/BUILD/root-6.28.00/core/unix/src/TUnixSystem.cxx:2104
#5  TUnixSystem::StackTrace (this=<optimized out>) at /builddir/build/BUILD/root-6.28.00/core/unix/src/TUnixSystem.cxx:2395
#6  0xf6f77e71 in DefaultErrorHandler (level=<optimized out>, abort_bool=<optimized out>, location=<optimized out>, msg=<optimized out>) at /builddir/build/BUILD/root-6.28.00/core/base/src/TErrorDefaultHandler.cxx:177
#7  0xf70489cc in ErrorHandler (level=6000, location=0xf7eba607 "", fmt=0xf71600ac "%s violated at line %d of `%s'", ap=0xffde8958 "\304\305\352\367R") at /builddir/build/BUILD/root-6.28.00/core/foundation/src/TError.cxx:153
#8  0xf704919a in Fatal (location=0xf7eba607 "", fmt=0xf71600ac "%s violated at line %d of `%s'") at /builddir/build/BUILD/root-6.28.00/core/foundation/src/TError.cxx:251
#9  0xf7e04248 in ROOT::Experimental::Detail::RColumn::Flush (this=0x652beb90) at /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RColumn.cxx:82
#10 0xf7e0a18d in ROOT::Experimental::Detail::RFieldBase::Flush (this=0xda3d4040) at /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RField.cxx:370
#11 0xf7e2c3e0 in ROOT::Experimental::RNTupleWriter::CommitCluster (this=0xd7486120, commitClusterGroup=true) at /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTuple.cxx:370
#12 0xf7e2c674 in ROOT::Experimental::RNTupleWriter::~RNTupleWriter (this=0xd7486120, __in_chrg=<optimized out>) at /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/src/RNTuple.cxx:327
#13 0x56560f58 in std::default_delete<ROOT::Experimental::RNTupleWriter>::operator() (__ptr=0xd7486120, this=<optimized out>) at /usr/include/c++/13/bits/unique_ptr.h:100
#14 0x565606cf in std::default_delete<ROOT::Experimental::RNTupleWriter>::operator() (__ptr=<optimized out>, this=<optimized out>) at /usr/include/c++/13/bits/unique_ptr.h:94
#15 std::unique_ptr<ROOT::Experimental::RNTupleWriter, std::default_delete<ROOT::Experimental::RNTupleWriter> >::~unique_ptr (this=0xffde8b50, __in_chrg=<optimized out>) at /usr/include/c++/13/bits/unique_ptr.h:405
#16 RNTuple_LargeFile2_Test::TestBody (this=0x57246b80) at /builddir/build/BUILD/root-6.28.00/tree/ntuple/v7/test/ntuple_extended.cxx:260
#17 0xf7d7159c in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [clone .constprop.0] () from /lib/libgtest.so.1.13.0
#18 0xf7d5cf46 in testing::Test::Run() () from /lib/libgtest.so.1.13.0
#19 0xf7d5d17a in testing::TestInfo::Run() () from /lib/libgtest.so.1.13.0
#20 0xf7d5d327 in testing::TestSuite::Run() () from /lib/libgtest.so.1.13.0
#21 0xf7d6a95b in testing::internal::UnitTestImpl::RunAllTests() () from /lib/libgtest.so.1.13.0
#22 0xf7d695cb in testing::UnitTest::Run() () from /lib/libgtest.so.1.13.0
#23 0xf7d1e137 in main () from /lib/libgtest_main.so.1.13.0
#24 0xf6837969 in __libc_start_call_main () from /lib/libc.so.6
#25 0xf6837a2c in __libc_start_main_impl () from /lib/libc.so.6
#26 0x56560e0b in _start ()
CMake Error at /builddir/build/BUILD/root-6.28.00/cmake/modules/RootTestDriver.cmake:232 (message):
  error code: Subprocess aborted
~~~

### PR DESCRIPTION
**Pull request by jblomer**
Fixes #12427

### CODE REVIEW

**Review by jalopezg-git — APPROVED** [review]

**Review by eguiraud — APPROVED** [review]
LGTM :)

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 1]
```diff
@@ -270,8 +270,9 @@ public:
    virtual std::vector<RFieldValue> SplitValue(const RFieldValue &value) const;
    /// The number of bytes taken by a value of the appropriate type
    virtual size_t GetValueSize() const = 0;
-   /// For many types, the alignment requirement is equal to the size; otherwise override.
-   virtual size_t GetAlignment() const { return GetValueSize(); }
+   /// As a rule of thumb, the alignment is equal to the size of the type. There are, however, various exceptions
+   /// to this rule depending on OS and CPU architecture. So enforce the alignment to be explicitly spelled out.
```
Comment (pcanal) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:274]: It seems to be declared but not use yet, is that the intent?

Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:274]: Not sure I understand.  Do you mean `GetAlignment()`? This is overwritten in the descendant classes.

**Comment by ellert** [comment]
This PR fixes the broken test on ix32. Consider backport to 6.28.

**Comment by jblomer** [comment]
Many thanks for confirming!  We are not yet aligning RNTuple developments with releases. Early testers / adopters are encouraged to follow the developments in the master branch.

**Pull request merged**